ibis/expr/analysis.py

```diff
@@ -1039,75 +1037,71 @@ def is_scalar_reduction(expr):
-        self.predicates: MutableSequence[ir.BooleanValue] = []
+    return op.to_expr()
 
-    def get_result(self) -> ir.BooleanColumn:
-        self._visit(self.expr)
 
-        op = self.expr.op()
+class _AnyToExistsTransform:
+    __slots__ = "expr", "tables", "predicates"
 
-        if isinstance(op, ops.Any):
-            op_class = ops.ExistsSubquery
-        else:
-            assert isinstance(op, ops.NotAny)
-            op_class = ops.NotExistsSubquery
+    def __init__(self, expr: ir.Expr) -> None:
+        self.expr = expr
+        self.tables: MutableSet[ir.Table] = set(expr.op().root_tables())
+        self.predicates: MutableSequence[ir.BooleanValue] = []
 
-        return op_class(self.foreign_table, self.predicates).to_expr()
+    def get_result(self) -> tuple[Sequence[ir.Table], list[ir.BooleanColumn]]:
+        self._visit(self.expr)
+        return (
+            tuple(table.to_expr() for table in self.tables),
+            self.predicates,
+        )
 
     def _visit(self, expr: ir.Expr) -> None:
-        node = expr.op()
-
+        if isinstance(expr, ir.Table):
+            self._visit_table_node(expr.op())
+        elif isinstance(expr, ir.BooleanColumn):
+            for sub_expr in flatten_predicate(expr):
+                self.predicates.append(sub_expr)
+                self._visit_node(sub_expr.op())
+        elif isinstance(expr, ir.Expr):
+            self._visit_node(expr.op())
+
+    def _visit_node(self, node: ops.Node) -> None:
         for arg in node.flat_args():
-            if isinstance(arg, ir.Table):
-                self._visit_table(arg)
-            elif isinstance(arg, ir.BooleanColumn):
-                for sub_expr in flatten_predicate(arg):
-                    self.predicates.append(sub_expr)
-                    self._visit(sub_expr)
-            elif isinstance(arg, ir.Expr):
-                self._visit(arg)
-
-    def _find_blocking_table(self, expr: ir.Expr) -> ir.Expr:
-        node = expr.op()
+            self._visit(arg)
 
-        if node.blocks():
-            return expr
+    def _visit_table_node(self, table: ops.TableNode) -> None:
+        assert isinstance(table, ops.TableNode)
 
-        return next(
-            (
-                result
-                for arg in node.flat_args()
-                if (
-                    isinstance(arg, ir.Expr)
-                    and (result := self._find_blocking_table(arg)) is not None
-                )
-            ),
-            None,
-        )
+        if self._find_blocking_table(table):
+            self.tables.add(table)
 
-    def _visit_table(self, expr: ir.Expr) -> None:
-        if (
-            isinstance(expr, ir.Table)
-            and (base_table := self._find_blocking_table(expr)) is not None
-            and base_table.op() not in self.query_roots
-        ):
-            self.foreign_table = expr
+    def _find_blocking_table(self, node: ops.TableNode) -> ops.TableNode:
```

Comment:
It seems to return with a `bool`.

Could you replace it with a new analysis function instead:

```python
def has_blocking_table(expr):
    def predicate(expr):
        op = expr.op()
        if op.blocks():
            return lin.halt, True
        else:
            return lin.proceed, None
        
    return any(lin.traverse(predicate, expr))
```

?

Reply:
For example, `util.to_op_graph` :)

Reply:
Yep, this can be replaced with a simple DAG iteration.

Reply:
Done.